File: problem1/sum_to_n.js
/*
Problem 1: Three ways to sum to n using javascript
Author: Anh Nguyen
Date: 14/02/2024
*/

// Iterative approach with complexity: O(n)
function sum_to_n_a(n) {
  let sum = 0;
  for (let i = 1; i <= n; i++) {
    sum += i;
  }
  return sum;
}

// Mathematical approach (most efficient) with complexity: O(1)
function sum_to_n_b(n) {
  return (n * (n + 1)) / 2;
}

// Recursive approach with complexity: O(n)
function sum_to_n_c(n) {
  if (n === 1) {
    return 1;
  }
  return n + sum_to_n_c(n - 1);
}

// Testing the functions
console.log(sum_to_n_a(15));
console.log(sum_to_n_b(15));
console.log(sum_to_n_c(15));

File: problem4/sum_to_n.go
/*
Problem 1: Three ways to sum to n using Golang
Author: Anh Nguyen
Date: 14/02/2024
*/

package main

import "fmt"

func main() {
	// Testing the functions
	fmt.Println(sum_to_n_a(680))
	fmt.Println(sum_to_n_b(680))
	fmt.Println(sum_to_n_c(680))

}

func sum_to_n_a(n int) int {
	//Mathematical approach with complexity O(1)
	return n * (n + 1) / 2
}

func sum_to_n_b(n int) int {
	//Using loop approach with complexity: O(n)
	sum := 0
	for i := 1; i <= n; i++ {
		sum += i
	}
	return sum
}

func sum_to_n_c(n int) int {
	//Recursive approach with complexity: O(n)
	if n == 1 {
		return 1
	}
	return n + sum_to_n_c(n-1)
}
